service/history/activityReplicator.go

```diff
@@ -0,0 +1,239 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination activityReplicator_mock.go
+
+package history
+
+import (
+	ctx "context"
+	"time"
+
+	"go.uber.org/cadence/.gen/go/shared"
+
+	h "github.com/uber/cadence/.gen/go/history"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	activityReplicator interface {
+		SyncActivity(
+			ctx ctx.Context,
+			request *h.SyncActivityRequest,
+		) error
+	}
+
+	activityReplicatorImpl struct {
+		historyCache    *historyCache
+		clusterMetadata cluster.Metadata
+		logger          log.Logger
+	}
+)
+
+func newActivityReplicator(
+	shard ShardContext,
+	historyCache *historyCache,
+	logger log.Logger,
+) *activityReplicatorImpl {
+
+	return &activityReplicatorImpl{
+		historyCache:    historyCache,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		logger:          logger.WithTags(tag.ComponentHistoryReplicator),
+	}
+}
+
+func (r *activityReplicatorImpl) SyncActivity(
+	ctx ctx.Context,
+	request *h.SyncActivityRequest,
+) (retError error) {
+
+	// sync activity info will only be sent from active side, when
+	// 1. activity has retry policy and activity got started
+	// 2. activity heart beat
+	// no sync activity task will be sent when active side fail / timeout activity,
+	// since standby side does not have activity retry timer
+
+	domainID := request.GetDomainId()
+	execution := workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowId,
+		RunId:      request.RunId,
+	}
+
+	context, release, err := r.historyCache.getOrCreateWorkflowExecution(ctx, domainID, execution)
+	if err != nil {
+		// for get workflow execution context, with valid run id
+		// err will not be of type EntityNotExistsError
+		return err
+	}
+	defer func() { release(retError) }()
+
+	msBuilder, err := context.loadWorkflowExecution()
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); !ok {
+			return err
+		}
+
+		// this can happen if the workflow start event and this sync activity task are out of order
+		// or the target workflow is long gone
+		// the safe solution to this is to throw away the sync activity task
+		// or otherwise, worker attempt will exceeds limit and put this message to DLQ
+		return nil
+	}
+
+	if !msBuilder.IsWorkflowExecutionRunning() {
```

Comment:
beginning from here 2DC logic and NDC logic should be separated.

NDC should first check whether the target workflow's branch match incoming version history, by using version history, not next event ID & version